### ISSUE TITLE
BaseStreamWriter.writeSpace(String) should not close open element

### ISSUE THREAD

**Issue by kwin** [FasterXML/woodstox#95]
Sometimes it is useful to also indent attributes within an element. For that `org.codehaus.stag2.writeSpace(String)` should be used. Unfortunately that will always close any open start tags so it cannot be used to indent attributes. As the javadoc is clearly stating this should be used for indentation (http://fasterxml.github.io/stax2-api/javadoc/3.1.4/org/codehaus/stax2/XMLStreamWriter2.html#writeSpace(java.lang.String)) it is unexpected that this behaves like `writeRaw(String)`

**Comment by cowtowncoder** [comment]
Hmmh. Indentation mentioned was meant for element indentation, and not for changing possible white-space between attributes. And `writeSpace()` specifically would only produce cdata sections, which inter-attribute white space would not be.
So this would be usage different from what was intended originally.
Use of `writeRaw()` however should be supported for such case.

But if you or anyone is interested in extending things to work for this usage as well, I think that would make sense.

**Comment by kwin** [comment]
> Use of writeRaw() however should be supported for such case.
Unfortunately not. If I call the following methods, I get the exception from above
```writeStartTag(...)
writeRaw(...)
writeAttribute(...)```

This is due to the fact that `writeRaw(...)` closes the start tag implicitly and therefore afterwards does no longer allow attributes to be added. The only workaround right now is to use the underlying writer directly to write the attribute indent (https://github.com/FasterXML/woodstox/blob/fee31c2efae3858635a2af212dcad09b250a84be/src/main/java/com/ctc/wstx/api/WstxOutputProperties.java#L184).

That feels wrong and IMHO the javadoc should state this limitation more explicitly in `writeSpace(...)` or in the best case this limitation should be lifted.

**Comment by cowtowncoder** [comment]
We are always open for patches: `writeRaw()` I think should not modify the state -- and should be usable for this usage -- for sure. Whether `writeSpace()` should work that way or not I don't have strong opinion, but would accept a patch to change too. Similarly for javadoc changes.

I don't have time to work on this myself in near-term but I think this could be something easy for others to pick up so will label as such.

**Comment by kwin** [comment]
> writeRaw() I think should not modify the state
This is tricky from a backwards-compatibility point of view as the API javadoc explicity says
>  It will not update state of the writer (except by possibly flushing output of previous writes, like finishing a start element)
(http://fasterxml.github.io/stax2-api/javadoc/3.1.4/org/codehaus/stax2/XMLStreamWriter2.html#writeRaw(java.lang.String))
Maybe adding an explicit flag whether start tags should be closed could be added as overloaded `writeRaw()` method.

**Comment by cowtowncoder** [comment]
Good point; it is bit of ambiguous case since without flushing like that possibly existing use cases could break.

Not sure how to proceed here, however, since as a practical matter I do not think I want to start making changes to Stax2 API just for this use case.

A likely workaround here might just be to instead use underlying `Writer` but first flushing `XMLStreamWriter`.

**Comment by kwin** [comment]
> ... without flushing like that possibly existing use cases could break.

That happened now with 7.2.0: #290

**Comment by cowtowncoder** [comment]
@kwin On original use case: would `writeSpace()` have worked instead of `writeRaw()`? That is:

```
writeStartTag(...)
writeSpace(...)
writeAttribute(...)
```

if so, I may have been too quick to do what was requested here.

**Comment by kwin** [comment]
In my tests `writeSpace()` behaves exactly like `writeRaw()`: https://github.com/FasterXML/woodstox/blob/336ceb874dadfd2ba135b860f9c4b3b2769745e9/src/main/java/com/ctc/wstx/sw/BaseStreamWriter.java#L1203-L1219

**Comment by cowtowncoder** [comment]
Ok. That is unfortunate. So this means that `writeSpace()` behavior changed in 7.2.0 as well.

So rollback would not really solve issue; and separating behavior would further confuse behavior (plus ideally avoided in patch anyway).

Not quite sure how to proceed here.

One good thing is that the work-around is harmless for pre-7.2.0 versions so at least it is possible to modify things to work same way in before and after versions.

**Comment by cowtowncoder** [comment]
I think this needs to be reverted in 7.2.1, due to widespread issues downstream.

And then add `writeRaw()` overload in Stax2 API (4.4.0), implement in Woodstox 7.3.

**Comment by kwin** [comment]
Can you reopen this issue then?

**Comment by cowtowncoder** [comment]
Let's not re-open this since it is referred to by release notes and references get complicated with re-opens, version included etc: instead, would you mind filing a new issue? Same/similar info, reference to this issue as background. Just a procedural thing but I've had problems in the past.

**Comment by kwin** [comment]
Done in #298
